### ISSUE TITLE
ty infers `int` in type annotation as `int` instead of `type[int]`

### ISSUE THREAD

**Issue by MichaReiser** [astral-sh/ty#3070]
Hovering over the `int` annotation of `param` shows `int`, whereas `reveal_type(int)` shows `class<int>`. hovering `param` correctly shows that it is an int.

[Playground](https://play.ty.dev/23e617d8-5dfe-4fda-8b0e-a9e4afa7fc1d)

`main.py`

```py
def foo(param: int):
    reveal_type(param)
    reveal_type(int)
```

Inferring `int` as `int` in type annotation positions is an issue for editors using ty as a type engine because it can result in false positives:

> If they’re specific to the new LSP request or when using its results

They are only visible when using the new request, as far as I know, because of how PyCharm is using it.

Here's an example of a false-positive error that is reported when using `ty` as a type engine
<img width="588" height="119" alt="image" src="https://github.com/user-attachments/assets/6a13b3e4-c9fc-404c-b08a-feb5bd8f7075" />

It happens because `ty` infers the type of `int` in the method signature as `int,` not `type[int]`. 
<img width="588" height="119" alt="image" src="https://github.com/user-attachments/assets/f4f0fb2b-f1f7-48ce-a256-a9f56d85e2e2" />

And `|` operator expects there to be `int` on the right hand side, not `None`. `type[int]`, on the other hand, would accept `None` on the right hand side for `|`

I couldn't reproduce this issue in `ty` itself, since doing a simple `reveal_type(int)` prints `type[int]`. The issue is specific to inferring the type of the annotation in the method signature

_Originally posted by @missingdays in https://github.com/astral-sh/ruff/issues/23836#issuecomment-4075793476_

**Comment by sharkdp** [comment]
> Hovering over the `int` annotation of `param` shows `int`, whereas `reveal_type(int)` shows `class<int>`.

(As you are probably aware), this is intentional. The `int` in `param: int` appears in a type-expression context. The `int` in `reveal_type(int)` appears in a value-expression context. In a type expression context, `int` represents a type that contains objects that are *instances* of the class `int`. In a value expression context, `int` represents the class `int` itself. And that class object is of type `<class 'int'>` (a type that only has one inhabitant, the class `int`).

So when we have something like `param: int` and hover over `int`, there are really two questions you could ask:
- Given that this is a symbol in type-expression context, what kind of type does this symbol *represent* in this type annotation?
- What is the type of this symbol itself *if we would encounter it in a value expression context*?

The answer to the first question is `int`, the answer to the second question is `<class 'int'>`. In the normal human-uses-cursor-to-hover-over-type-annotation-symbol use case, the first question is much more relevant. So I like what the ty LSP currently does.

But I understand that there might be use cases where you want an answer to that second question. Arguably, that should be a different request type though.

Note that ty has type-level operations (`Type::to_meta_type` / `Type::to_instance`) that can translate between these two layers. However, those operations are lossy in some cases.

**Comment by MichaReiser** [comment]
> (As you are probably aware), this is intentional.

You're giving me too much credit 😆 

> The answer to the first question is int, the answer to the second question is <class 'int'>. In the normal human-uses-cursor-to-hover-over-type-annotation-symbol use case, the first question is much more relevant. So I like what the ty LSP currently does.

That makes sense to me. Hover might not have been the best example, because this is more of an API design question: Is it expected that `inferred_type` on `int` in a type annotation position returns `int` over `type[int]`? 


I don't have a good sense for this, and today's behavior is very reasonable, because it makes hover work without any special casing. However, I wonder whether this will become more of a problem if we build lint rules outside `TypeInferenceBuilder` that use `inferred_type`? Do we expect those rules to be aware whether they're inside a type-context (which isn't always easy to determine, thanks to type aliases) or not and, because of it, they need to interpret the inferred type differently. 

> But I understand that there might be use cases where you want an answer to that second question. Arguably, that should be a different request type though.


We already use a different request in https://github.com/astral-sh/ruff/pull/23836, the problem is that it's based on `inferred_type`, which returns `int`.

**Comment by carljm** [comment]
There may not be an answer that satisfies all use cases well here, short of storing up-to-two types per expression.
